File: python/61_Rotate_List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head:
            return head
        length = 0
        tail, curr = None, head
        while curr:
            length += 1
            tail, curr = curr, curr.next
        tail.next = head
        for _ in range(0, length - k % length):
            tail, head = head, head.next
        tail.next = None
        return head

File: c++/61_Rotate_List/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head)
            return head;
        int len = 0;
        ListNode * tail = nullptr;
        ListNode * curr = head;
        while (curr) {
            len++;
            tail = curr;
            curr = curr->next;
        }
        k = k % len;
        int count = len - k;
        tail->next = head;
        ListNode * prev = tail;
        while (count > 0) {
            count -= 1;
            prev = head;
            head = head->next;
        }
        prev->next = nullptr;
        return head;
    }
};
